### DISCUSSION TITLE
[Feature Request] Allow GitHub apps to be created privately across multiple orgs in an Enterprise

### DISCUSSION THREAD

**Discussion by jahredhope** [community/community#7439] [Enterprise]
Currently to create a GitHub App across multiple orgs you need to make the GitHub App public or duplicate the App.

For Enterprise Cloud users that have multiple orgs this means exposing internal behaviours (the names and descriptions of GitHub apps) publicly.

Is it on the roadmap to allow GitHub Apps a 3rd visibility option, between Private and Public? One that would allow users make a GitHub App private across an Enterprise, or an explicit list of orgs?

**Comment by jorsmatthys** [comment]
Hello, we also want this feature at our enterprise. 

Currently we have apps that we list as public, so that we can install them on multiple (in our case 8) organizations. 
There are other settings at  the enterprise level that prevent the public URL of our apps to be visible to people outside of the organization, but this still feels like a hack. 
It feels like being able to mark GitHub Apps as "internal" would be a logical addition.

**Reply by leostolyarov** [reply]
+1

A really important feature to have for enterprises with multiple repositories.

**Reply by fblacuttgottret-godaddy** [reply]
+1

We need this and I would expect this functionality to be part of the enterprise feature set.

**Reply by xino12** [reply]
+1

**Reply by nuugen** [reply]
+1

**Reply by HarshitNagar** [reply]
> There are other settings at the enterprise level that prevent the public URL of our apps to be visible to people outside of the organization

@jorsmatthys what are these settings ? Do they also prevent access to the public URL or just hide it ?

**Reply by tyleris-irish** [reply]
+1, surprised this isn't a thing. My current plan is to create the same app across 5 orgs under our Enterprise Cloud, but this will be an unnecessary pain because I'm dealing with 5 different private keys.

**Comment by lavoiesl** [comment]
This seems to be a duplicate of https://github.com/orgs/community/discussions/24549

**Reply by jahredhope** [reply]
I don't think so.

That issue appears to be about whether you can have an app's tokens access two repositories across organisations at once.

This issue is about the app itself being private or public, i.e. whether public users can see the app. Where currently, you need to make an app public to install it on multiple organisations within an enterprise.

**Comment by ina-schaal** [comment]
Hi there!
We also need this feature for our enterprise account. Is there a timeline for this? 
Best regards!

**Comment by equalizer999** [comment]
We are currently working on such a feature as mentioned in our [Roadmap](https://github.com/orgs/github/projects/4247/views/1?filterQuery=is%3Aopen+-status%3A%22Q4+2022+%E2%80%93+Oct-Dec%22%2C%22Q2+2023+%E2%80%93+Apr-Jun%22%2C%22Q3+2023+%E2%80%93+Jul-Sep%22%2C%22Q4+2023+%E2%80%93+Oct-Dec%22+enterprise+app&pane=issue&itemId=65252371)

**Reply by tenjaa** [reply]
Can you clarify the last part?

> The initial permission that this will ship with is the ability to manage GitHub Apps installed on organizations in the enterprise. This will allow administrators to automate the installation, uninstallation, and blocking of apps within their enterprise. Note, though, that this is an enterprise-level access still and "Enterprise apps" don't otherwise get access to organizations in the enterprise.

The first two sentences sound to me, that I can say on the enterprise level "install App X on all orgs in my enterprise".
The last sentence sounds like that is for enterprise level only and "App X" does not get access to the org? But what else can it do then in the org, if not accessing the orgs data?

**Reply by GreyBouton** [reply]
I talked to corporate. Leadership says we are going to need this ASAP (in the next 12 hours) so please prioritize this. On behalf of the UN, thank you for your cooperation.

**Reply by ChooJosh** [reply]
Please have this done as high priority.

**Reply by procommand** [reply]
I'd like this done as a high priority, this is very crucial to a project being led in Afghanistans embassy and their need for this is essential to their success as a leadership and being able to comply with other nations. Please rush this

**Reply by procommand** [reply]
@joe-tingsanchali-sonarsource and @jamietanna do we have a problem we need to discuss? Keep your low priority opinion to yourself, clearly you are the least impacted by this and least important in this project.

**Reply by queenofcorgis** [reply]
@procommand I understand that you are excited and interested in this feature, but with a community of folks from around the world working with GitHub in a unique way folks may respectfully disagree with what needs to be prioritized. But remember to criticize ideas, not people. Be courteous to those you interact with here in the Community. Please be aware that breaking our standards is considered a violation of our [Code of Conduct](https://github.com/community/community/blob/fd6011231047419e2a4b1ac73fbb6aec10d86c38/CODE_OF_CONDUCT.md) and our moderator team may take appropriate action.

**Comment by hpsin** [comment]
Hey folks - this is in final testing right now, to make sure it's working securely, with the goal of releasing it before Universe (i.e. in the next week and a half). Specifically this is the ability for an enterprise to own an application and make it available to just their organizations and members. Keep an eye out here and the changelog for details.

**Comment by hpsin** [comment]
As promised - here we are! Feedback welcome
https://github.blog/changelog/2024-10-22-enterprises-can-create-github-apps-for-use-within-the-enterprise/

**Reply by hpsin** [reply]
This is now GA, and supports transferring an app to the enterprise from a user or org: https://github.blog/changelog/2025-03-10-enterprise-owned-github-apps-are-now-generally-available/

**Reply by jamietanna** [reply]
Hey!

Re:

> In addition, permission updates made to an enterprise-owned App are now automatically accepted by all of the organizations in the enterprise.

Does that mean an App Manager on an Enterprise-owned App can change permissions and they'll be auto-approved for any Org in the Enterprise, without a Org Owner approving them?

**Reply by hpsin** [reply]
That's correct. Right now only enterprise owners can manage the enterprise-owned apps, but we're working on adding app managers as well for enterprise-owned apps. The logic is that the enterprise owner can simply add themselves to the org and approve the update. (and clearly wants the permissions to be updated). This has the potential for impact if somehow a developer/admin was reliant on the app _not_ being updated in a specific organization. 

We are currently debating whether this auto-update behavior should extend to someone that is trusted to be an app manager (for the enterprise or for a specific app), as it can be used to effectively escalate privilege within every organization where the app is installed. It's likely that this privilege will remain with the enterprise admin as well as a new fine-grained permission like `Can update enterprise app permissions and have them automatically accepted`, which puts that control back in the hands of the admin who would grant that permission to a user.

**Comment by Cavaler** [comment]
This still does not allow you to create a cross-organization token, or does it?

**Reply by hpsin** [reply]
That's correct - this is _just_ ownership. 

Installing an app onto the enterprise is currently in testing, which is this item: https://github.com/github/roadmap/issues/1045 - this lets you target _just_ the enterprise account (not the orgs or repos within it) with a GitHub App. 

After this item we would look at giving that enterprise installation access to _all_ of the orgs and repos in the enterprise, so you can use one token to leverage the app's permissions everywhere in the enterprise.

**Reply by yongzhang** [reply]
Hi @hpsin Any progress on the cross-org github token plz? Currently we have to create 2 separete github app client based on the installation id to pull private repos which is very annoying.

**Reply by evilhamsterman** [reply]
@yongzhang it's been available for almost a year https://docs.github.com/en/enterprise-cloud@latest/admin/managing-github-apps-for-your-enterprise/creating-github-apps-for-your-enterprise

**Reply by jamietanna** [reply]
I think they're referring to having an App token (or a fine-grained personal access token) that can access multiple orgs at once
